src/peft/tuners/delora/layer.py

```diff
@@ -0,0 +1,630 @@
+            self.delora_initial_w_norm[adapter_name] = w_norm.clone()
+
+        # initial copies and alpha (kept non-trainable). Stored on CPU only when use_residual_init=True
+        if init_weights is not False:
+            if self.delora_A[adapter_name].device.type != "meta":
+                if self.use_residual_init:
+                    # For real tensors, copy the actual initialized data
+                    self.delora_initial_A[adapter_name] = (self.delora_A[adapter_name].detach().to(torch.float32, copy=True, non_blocking=False).cpu())
+                    self.delora_initial_B[adapter_name] = (self.delora_B[adapter_name].detach().to(torch.float32, copy=True, non_blocking=False).cpu())
+                else:
+                    weight = self.get_base_layer().weight
+                    # For real tensors, copy to the target device
+                    self.delora_initial_A[adapter_name] = (self.delora_A[adapter_name].detach().to(device=weight.device, dtype=torch.float32, copy=True))
+                    self.delora_initial_B[adapter_name] = (self.delora_B[adapter_name].detach().to(device=weight.device, dtype=torch.float32, copy=True))
+                self.delora_initial_alpha[adapter_name] = torch.tensor([float(self.delora_alpha[adapter_name].detach().item())], dtype=torch.float32)
+            else:
+                # For meta tensors, create meta tensors for initial copies too
+                self.delora_initial_A[adapter_name] = torch.empty_like(self.delora_A[adapter_name], dtype=torch.float32, device="meta")
+                self.delora_initial_B[adapter_name] = torch.empty_like(self.delora_B[adapter_name], dtype=torch.float32, device="meta")
+                self.delora_initial_alpha[adapter_name] = torch.tensor([float(alpha)], dtype=torch.float32)
+
+            if self.use_residual_init and adapter_name not in self._initials_applied:
+                # store initial base weight norm
+                self.delora_initial_w_norm[adapter_name] = w_norm
+                # subtract initial delta using the reference norm (only if not on meta device)
+                if self.delora_initial_A[adapter_name].device.type != "meta" and w.device.type != "meta":
+                    with torch.no_grad():
+                        # subtract exact initial delta so effective initial effect is zero
+                        A0 = self.delora_initial_A[adapter_name].to(dtype=w.dtype, device=w.device)
+                        B0 = self.delora_initial_B[adapter_name].to(dtype=w.dtype, device=w.device)
+                        alpha0 = self.delora_initial_alpha[adapter_name].to(dtype=w.dtype, device=w.device).squeeze(0)
+                        wnorm0 = self.delora_initial_w_norm[adapter_name].to(dtype=w.dtype, device=w.device)
+                        delta0 = self._compute_delta_from_params(A0, B0, alpha0, self.r[adapter_name], wnorm0)
+                        self.get_base_layer().weight.data.sub_(delta0)
```

Comment:
If we add two DeLoRA adapters with `use_residual_init=True` to the model, would the 2nd one not subtract its delta again from the base weight? Would this not result in the base weight being incorrect for the first adapter?